File: day-backend/app/presentation/schemas/analytics.py
"""Analytics Pydantic schemas for request/response models."""

from __future__ import annotations

import uuid
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.domain.analytics.value_objects import Granularity, PeriodPreset

# ---------- Request Schemas ----------


class AnalyticsFilters(BaseModel):
    """Query parameters for analytics endpoints."""

    date_from: date | None = None
    date_to: date | None = None
    period: PeriodPreset | None = None
    granularity: Granularity = Granularity.DAY
    property_id: list[uuid.UUID] | None = None
    source: str | None = None


# ---------- Response Schemas ----------


class PropertyMetricsResponse(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
        json_encoders={Decimal: float},
    )

    property_id: uuid.UUID
    property_name: str
    property_internal_name: str
    revenue: Decimal
    adr: Decimal
    revpar: Decimal
    expenses: Decimal
    profit: Decimal
    commission: Decimal
    vacancy_days: int
    occupancy_rate: Decimal
    avg_stay_duration: Decimal
    total_bookings: int
    booked_nights: int


class AnalyticsSummaryResponse(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
        json_encoders={Decimal: float},
    )

    total_revenue: Decimal
    total_expenses: Decimal
    total_profit: Decimal
    total_commission: Decimal
    overall_adr: Decimal
    overall_revpar: Decimal
    overall_occupancy_rate: Decimal
    avg_stay_duration: Decimal
    total_bookings: int
    total_booked_nights: int
    total_vacancy_days: int
    properties_count: int


class AnalyticsMetricsResponse(BaseModel):
    """Full analytics response: summary + per-property breakdown."""

    summary: AnalyticsSummaryResponse
    properties: list[PropertyMetricsResponse]
    date_from: date
    date_to: date


class TimeSeriesPointResponse(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
        json_encoders={Decimal: float},
    )

    period_start: date | None = None
    period_label: str
    revenue: Decimal
    bookings_count: int
    booked_nights: int
    occupancy_rate: Decimal


class TimeSeriesResponse(BaseModel):
    """Time series analytics response."""

    data: list[TimeSeriesPointResponse]
    granularity: Granularity
    date_from: date
    date_to: date


class ExportResponse(BaseModel):
    """CSV export response metadata."""

    filename: str
    content_type: str = "text/csv"

File: day-frontend/src/types/analytics.ts
export type PeriodPreset = 'week' | 'month' | 'quarter' | 'year' | 'custom'
export type Granularity = 'day' | 'week' | 'month'

export interface PropertyMetrics {
  property_id: string
  property_name: string
  property_internal_name: string
  revenue: number
  adr: number
  revpar: number
  expenses: number
  profit: number
  commission: number
  vacancy_days: number
  occupancy_rate: number
  avg_stay_duration: number
  total_bookings: number
  booked_nights: number
}

export interface AnalyticsSummary {
  total_revenue: number
  total_expenses: number
  total_profit: number
  total_commission: number
  overall_adr: number
  overall_revpar: number
  overall_occupancy_rate: number
  avg_stay_duration: number
  total_bookings: number
  total_booked_nights: number
  total_vacancy_days: number
  properties_count: number
}

export interface AnalyticsMetricsResponse {
  summary: AnalyticsSummary
  properties: PropertyMetrics[]
  date_from: string
  date_to: string
}

export interface TimeSeriesPoint {
  period_start: string | null
  period_label: string
  revenue: number
  bookings_count: number
  booked_nights: number
  occupancy_rate: number
}

export interface TimeSeriesResponse {
  data: TimeSeriesPoint[]
  granularity: Granularity
  date_from: string
  date_to: string
}

export interface AnalyticsFilters {
  date_from?: string
  date_to?: string
  period?: PeriodPreset
  granularity?: Granularity
  property_ids?: string[]
  source?: string
}
